src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs

```diff
@@ -234,7 +256,33 @@ Checksum IChecksummedObject.Checksum
                 {
                     if (_lazyChecksum == null)
                     {
-                        _lazyChecksum = Checksum.Create(nameof(DocumentAttributes), this);
+                        using (var stream = SerializableBytes.CreateWritableStream())
+                        using (var writer = new ObjectWriter(stream))
+                        {
+                            writer.WriteString(nameof(DocumentAttributes));
+
+                            if (FilePath == null || ProjectFilePath == null)
+                            {
+                                // this checksum is not persistable because
+                                // this info doesn't have non volatile info
+                                Id.WriteTo(writer);
```

Comment:
> it just can't be persisted and re-used in different session.

The problem is that the *client* has no way to know if a checksum is persistable or not.  Thus, they cannot use checksums as the persistence key.  

I would prefer if there was either a boolean that says that the checksum needs to be persistable, or have two entrypoints "GetChecksum" and "GetPersistableChecksum".  The client can then ask for a persistable checksum and decide what to do if they can't get one.  

I get that we may not be able to make a "consistent, deterministic *persistable* checksum" in all cases.  What i'm asking for is that in those cases we don't *silently* fall over to making a non-persistable checksum.  

--

And, at the same time, i would really like us to actually figure out if we can make these persistable in some manner.